mongo/integration/unified/event_verification.go

```diff
@@ -12,28 +12,108 @@ import (
 	"fmt"
 
 	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/event"
+)
+
+var (
+	connectionClosedReasons = map[string]string{
+		"stale":      event.ReasonStale,
+		"idle":       event.ReasonIdle,
+		"error":      event.ReasonConnectionErrored,
+		"poolClosed": event.ReasonPoolClosed,
+	}
+
+	checkOutFailedReasons = map[string]string{
+		"poolClosed":      event.ReasonPoolClosed,
+		"timeout":         event.ReasonTimedOut,
+		"connectionError": event.ReasonConnectionErrored,
+	}
 )
 
 type commandMonitoringEvent struct {
 	CommandStartedEvent *struct {
 		Command      bson.Raw `bson:"command"`
 		CommandName  *string  `bson:"commandName"`
 		DatabaseName *string  `bson:"databaseName"`
+		HasServiceID *bool    `bson:"hasServiceId"`
 	} `bson:"commandStartedEvent"`
 
 	CommandSucceededEvent *struct {
-		CommandName *string  `bson:"commandName"`
-		Reply       bson.Raw `bson:"reply"`
+		CommandName  *string  `bson:"commandName"`
+		Reply        bson.Raw `bson:"reply"`
+		HasServiceID *bool    `bson:"hasServiceId"`
 	} `bson:"commandSucceededEvent"`
 
 	CommandFailedEvent *struct {
-		CommandName *string `bson:"commandName"`
+		CommandName  *string `bson:"commandName"`
+		HasServiceID *bool   `bson:"hasServiceId"`
 	} `bson:"commandFailedEvent"`
 }
 
+type cmapEvent struct {
+	ConnectionCreatedEvent *struct{} `bson:"connectionCreatedEvent"`
+
+	ConnectionReadyEvent *struct{} `bson:"connectionReadyEvent"`
+
+	ConnectionClosedEvent *struct {
+		Reason *string `bson:"reason"`
+	} `bson:"connectionClosedEvent"`
+
+	ConnectionCheckedOutEvent *struct{} `bson:"connectionCheckedOutEvent"`
+
+	ConnectionCheckOutFailedEvent *struct {
+		Reason *string `bson:"reason"`
+	} `bson:"connectionCheckOutFailedEvent"`
+
+	ConnectionCheckedInEvent *struct{} `bson:"connectionCheckedInEvent"`
+
+	PoolClearedEvent *struct {
+		HasServiceID *bool `bson:"hasServiceId"`
+	} `bson:"poolClearedEvent"`
+}
+
 type expectedEvents struct {
```

Comment:
The JSON structure for this type was modified in the related spec changes. Previously, the JSON file contained something like
```
{ client: "client0", events: [ <event0>, <event1>, ... ] }
```
To support both command and CMAP events, the new structure is
```
{ client: "client0", eventType: < either "command" or "cmap">, events: [ <event0>, <event1>, ... ] }
```

I opted to not include the new `eventType` field in this struct. Instead, I split up the `Events` slice into `CommandEvents` and `CMAPEvents` and wrote a custom BSON unmarshaller to examine the `eventType` value and unmarshal the JSON `events` array into once of these slices. The invariant is that exactly one slice will be non-nil once unmarshalling is complete.